quintoandar_kafka/kafka_consumer.py

```
@@ -3,53 +3,55 @@
 
 from retrying import retry
 from kafka.consumer.group import KafkaConsumer
+from quintoandar_kafka import IdempotenceClient
+from functools import partial
 
 
-class KafkaConsumerClient:
-
-    def __init__(self, group_id, bootstrap_servers, topic,
-                 processor, deserializer=None, idempotence_client=None):
+class KafkaDefaultConsumer():
+    def __init__(self, processor, **kwargs):
         self.log = logging.getLogger(__name__)
-        self.topic = topic
         self.processor = processor
-        self.deserializer = deserializer or self.default_deserializer
-        self.idempotence_client = idempotence_client or\
-            DefaultIdempotenceClient()
-        self.consumer = self.connect(group_id, bootstrap_servers)
-        self.consumer.subscribe(topic)
+        if kwargs.get('deserializer') is None:
+            kwargs['deserializer'] = self.default_deserializer
+        if kwargs.get('auto_offset_reset') is None:
+            kwargs['auto_offset_reset'] = 'latest'
+        self.consumer = self.connect(kwargs)
+        self.consumer.subscribe(kwargs['topic'])
 
     @retry(stop_max_attempt_number=10, wait_fixed=3000)
-    def connect(self, group_id, bootstrap_servers):
-        return KafkaConsumer(
-            value_deserializer=self.deserializer,
-            group_id=group_id,
-            bootstrap_servers=bootstrap_servers,
-            auto_offset_reset='latest')
+    def connect(self, kwargs):
+        return KafkaConsumer(kwargs)
+
+    def default_deserializer(self, m):
+        try:
+            return json.loads(m.decode('utf8'))
+        except Exception as ex:
+            self.log.error("Failed to decode message: %s", ex, exc_info=True)
```

You could use self.log.exception, which is the same as self.logerror with exc_info=True).

_But!_ I don't think we need to log the stacktrace! Because instead I think we should reraise the exception here, and leave it up to the client to catch it (and print the stack if he wants). It's generally a bad practice as a lib to catch an exception, because then there's no way the client can ever retrieve it. So:

```
except Exception as ex:
    self.log.error("Failed to decode message: %s", ex)
    raise
```

Actually I'm not even sure if we need to log anything. The underlying lib already logs a lot of stuff, and if we raise the exception it's up for the client to treat it as he sees fit. Might just want to use a lambda.

```
key_deserializer=lambda m: json.loads(m.decode('utf-8'))
```

---

It's true that the client could treat it. But the problem is the an exception would stop the event processing loop, effectively stopping the consumer.